### DISCUSSION TITLE
Change tables environment for latex output

### DISCUSSION THREAD

**Discussion by xomaxy** [quarto-dev/quarto-cli#2786] [Q&A] [answered]
Hi, I have created a cuarto document with this document class

`documentclass: scrartcl
  classoption: [twocolumn]
`

when I tried to do a table I get the following error

Package longtable Error: longtable not in 1-column mode.

I know that this happens because I cannot put a longtable in 2-colum mode. I want to change the environment to supertabular  or table*, but there is not an option in quarto to change the environment like figs with fig-env.

So how I put the table? I don't want to use the multicols package (I have already tried but the only solution is to scape the table of the multicol), why quarto doesn't have a tbl-env option?

**Comment by dragonstyle** [comment, marked as answer]
We haven't yet create a `table-env` option, though I agree that would be great. Here is a "workaround" I found and used in our Elsevier format, which redefines the longtable command and uses mini pages. Not 100% sure this will solve your problem, but though it might be useful:

https://github.com/quarto-journals/elsevier/blob/main/_extensions/elsevier/partials/_two-column-longtable.tex

We essentially just include this in the head when we detect a two column elseiver layout (using `include-in-header`).

**Reply by naisargi1101** [reply]
where to include this lines? Before lontable or where the pakages are defined?

**Reply by mcanouil** [reply]
As stated in the post, in the header using `include-in-header` (see <https://quarto.org/docs/output-formats/pdf-basics.html#latex-includes>) or by using/replacing partials, see <https://quarto.org/docs/journals/templates.html#template-partials>.

**Reply by checkdgt** [reply]
This worked perfectly for me with the same issue on the acm-pdf quarto template .. :)

**Reply by GregDMeyer** [reply]
I added that TeX snippet using `include-in-header` and I see that it is included in the produced TeX output, but I still am getting the same error (`Error: longtable not in 1-column mode.`). Any idea why this would be?

**Reply by mcanouil** [reply]
@GregDMeyer unfortunately, without a reproducible example, it's unlikely someone will have an "idea".

Could you share a **small self-contained "working" (reproducible)** example to work with, _i.e._, a complete Quarto document or a Git repository? The goal is to make it as easy as possible for us to recreate your problem so that we can fix it: please help us help you! Thanks.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

{{< lipsum 1 >}}

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image]({{< placeholder 600 400 >}}){#fig-placeholder}

{{< lipsum 1 >}}

The end after @fig-placeholder.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by GregDMeyer** [reply]
Ah, apologies for not being more specific! Here is a minimal (not) working example:

````qmd
---
format:
  pdf:
    documentclass: quantumarticle
    classoption:
    - noarxiv
    fontfamily: braket
    include-in-header:
     - file: table-fix.tex
---

| $a$ | $b$ | $c_\mathrm{out}$ |
|-----|-----|------------------|
| 0   | 0   | 0                |
| 0   | 1   | $c_\mathrm{in}$  |
| 1   | 0   | $c_\mathrm{in}$  |
| 1   | 1   | 1                |

: Relationship between input and output carries, for a single full adder with bit inputs $a$, $b$, and $c_in$. {#tbl-carries}
````

`table-fix.tex` contains the contents of the linked file in the thread.

It appears that the issue comes from the fact that I am using the `quantumarticle` class; I observe that if I use `article` and set the `twocolumn` option, it compiles successfully. I guess `quantumarticle` is just not well supported by quarto :-( (for example, this is why I needed to put the weird `fontfamily: braket` hack, because `quantumarticle` is not happy with the `lmodern` package included by default).

Anyway, the output of `quarto check`, since it was requested:

```
Quarto 1.5.54
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.2.0: OK
      Dart Sass version 1.70.0: OK
      Deno version 1.41.0: OK
      Typst version 0.11.0: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.5.54
      Path: /opt/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: (not installed)
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: Installation From Path
      Path: /usr/bin
      Version: undefined

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.12.4
      Path: /usr/bin/python3
      Jupyter: 5.7.2
      Kernels: qualtran, cirq, python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........(None)

      Unable to locate an installed version of R.
      Install R from https://cloud.r-project.org/

```

**Reply by GregDMeyer** [reply]
to be clear---I totally understand if the answer is simply "quarto and the quantumarticle class do not play well together." Just was hoping there is some solution.

**Reply by mcanouil** [reply]
The issue is not really Quarto but LaTeX which natively has a lot of conflicts that are in general very hard to avoid.

Here you have to debug the LaTeX part yourself using `keep-tex: true`. Try to render manually the Tex file and figure out what is causing the conflict/error.
Once done you need to figure out how to solve it.
Using `include-*` and template partials can help out setup the solution if there is one.

**Comment by xomaxy** [comment]
Thanks, I don't really understand all the latex code but it works! also I have found that you can scape the two columns mode with the `\onecolumn` `\twocolumn` command. Not ideal but it kind of works.

**Reply by cmarquardt** [reply]
I ran into a similar problem but I have difficulties getting it to work in my particular document template - and just pasting the code into my modified Pandoc template for my template also didn't work well so far. As far as I understand the code, it (a) wouldn't work without twocolumn being active and (b) only seems to support single-column tables, is that correct?

Many thanks.

**Comment by cmarquardt** [comment]
Hmm - to qualify my comment above a bit more - pasting the code from the solution in the Elsevier template before the `\begin{document}` didn't work in my case because parts of text placed *before* the markdown table then disappeared from the output generated by LaTeX. I didn't trace that down entirely because...

What did work was generating tables with calls to either `knitr::kable()` or `kableExtra::kbl()`; the advantage of these R functions is that they generate LaTeX (or HTML) output directly rather than going through Pandoc (and hence using a `longtable` environment). The main drawback is that cell options (like `label:` and `tbl-cap:`) cannot be used because they are not automatically passed on to the R function calls. And, of course, creating a data frame and then generating a table with `kable()` is more complicated than using plain markdown. An example for a one-column table is (to be placed in an R code block):

```{r}
knitr::kable(head(mtcars), caption = "Cars", label = "tab-cars", booktabs = TRUE)
```

while a full-width table can be generated with

```{r}
knitr::kable(head(mtcars), caption = "Cars", label = "tab-cars-wide", booktabs = TRUE, table.env = 'table*')
```

I do believe, though, that being able to set the table environment to `table` or `table*` would be beneficial if Pandoc allows that somehow.
